File: GameplayCore/Transform.cpp
#include "pch.h"
#include "Transform.h"

namespace EduEngine
{
	Transform::Transform(GameObject* parent) :
		Component(parent),
		m_Position{0, 0, 0},
		m_Rotation{0, 0, 0, 0},
		m_Scale{1, 1, 1},
		m_Dirty{ false }
	{
		UpdateWorldMatrix();
	}

	void Transform::Update()
	{
		if (m_Dirty)
		{
			UpdateWorldMatrix();
			m_Dirty = false;
		}
	}

	void Transform::SetPosition(Vector3 position)
	{
		m_Position = position;
		m_Dirty = true;
	}

	void Transform::SetRotation(Quaternion rotation)
	{
		m_Rotation = rotation;
		m_Dirty = true;
	}

	void Transform::SetScale(Vector3 scale)
	{
		m_Scale = scale;
		m_Dirty = true;
	}

	void Transform::UpdateWorldMatrix()
	{
		m_WorldMatrix = Matrix::CreateScale(m_Scale) *
			Matrix::CreateFromQuaternion(m_Rotation) *
			Matrix::CreateTranslation(m_Position);
	}
}

File: GameplayFramework/Transform.cs
using System.Numerics;

namespace EduEngine
{
    public class Transform : Component
    {
        private Vector3 _position;
        private Quaternion _rotation;
        private Vector3 _localScale;
        private bool _isDirty;

        public Transform(GameObject parent)
            : base(parent)
        {
            Position = Vector3.Zero;
            Rotation = Quaternion.Identity;
            LocalScale = Vector3.One;

            UpdateWorldMatrix();
            _isDirty = false;
        }

        public Vector3 Position
        {
            get => _position;
            set
            {
                _position = value;
                _isDirty = true;
            }
        }
        public Quaternion Rotation
        {
            get => _rotation;
            set
            {
                _rotation = value;
                _isDirty = true;
            }
        }
        public Vector3 LocalScale
        {
            get => _localScale;
            set
            {
                _localScale = value;
                _isDirty = true;
            }
        }
        public Vector3 Forward => Vector3.Transform(Vector3.UnitZ, Rotation);
        public Vector3 Right => Vector3.Transform(Vector3.UnitX, Rotation);
        public Vector3 Up => Vector3.Transform(Vector3.UnitY, Rotation);
        public Matrix4x4 WorldMatrix { get; private set; }

        public override void Update()
        {
            if (_isDirty == false)
                return;

            UpdateWorldMatrix();
        }

        public void SetWorldMatrix(Matrix4x4 matrix)
        {
            WorldMatrix = matrix;

            Matrix4x4.Decompose(matrix, out Vector3 scale, out Quaternion rotation, out Vector3 translation);
            Position = translation;
            Rotation = rotation;
            LocalScale = scale;
        }

        private void UpdateWorldMatrix()
        {
            WorldMatrix = Matrix4x4.CreateScale(LocalScale) *
                Matrix4x4.CreateFromQuaternion(Rotation) *
                Matrix4x4.CreateTranslation(Position);
        }
    }
}
